massa-grpc/src/api.rs

```diff
@@ -0,0 +1,894 @@
+//! Copyright (c) 2022 MASSA LABS <info@massa.net>
+//! gRPC API for a massa-node
+use std::{collections::HashMap, error::Error, io::ErrorKind, pin::Pin, str::FromStr};
+
+use crate::config::GrpcConfig;
+use itertools::izip;
+use massa_consensus_exports::{ConsensusChannels, ConsensusController};
+use massa_execution_exports::ExecutionController;
+use massa_models::{
+    address::Address,
+    block::{BlockDeserializer, BlockDeserializerArgs, SecureShareBlock},
+    endorsement::{EndorsementDeserializer, SecureShareEndorsement},
+    error::ModelsError,
+    operation::{OperationDeserializer, SecureShareOperation},
+    secure_share::SecureShareDeserializer,
+    slot::Slot,
+    timeslots,
+};
+use massa_pool_exports::{PoolChannels, PoolController};
+use massa_pos_exports::SelectorController;
+use massa_proto::massa::api::v1::{self as grpc, grpc_server::GrpcServer, FILE_DESCRIPTOR_SET};
+use massa_serialization::{DeserializeError, Deserializer};
+
+use futures_util::{FutureExt, StreamExt};
+use massa_protocol_exports::ProtocolCommandSender;
+use massa_storage::Storage;
+use tokio::sync::oneshot;
+use tonic::codegen::{futures_core, CompressionEncoding};
+use tonic_web::GrpcWebLayer;
+use tracing::log::{error, info, warn};
+
+/// gRPC API content
+pub struct MassaService {
+    /// link(channels) to the consensus component
+    pub consensus_controller: Box<dyn ConsensusController>,
+    /// link(channels) to the consensus component
+    pub consensus_channels: ConsensusChannels,
+    /// link to the execution component
+    pub execution_controller: Box<dyn ExecutionController>,
+    /// link(channels) to the pool component
+    pub pool_channels: PoolChannels,
+    /// link to the pool component
+    pub pool_command_sender: Box<dyn PoolController>,
+    /// link(channels) to the protocol component
+    pub protocol_command_sender: ProtocolCommandSender,
+    /// link to the selector component
+    pub selector_controller: Box<dyn SelectorController>,
+    /// link to the storage component
+    pub storage: Storage,
+    /// gRPC configuration
+    pub grpc_config: GrpcConfig,
+    /// node version
+    pub version: massa_models::version::Version,
+}
+
+impl MassaService {
+    /// Start the gRPC API
+    pub async fn serve(
+        self,
+        config: &GrpcConfig,
+    ) -> Result<StopHandle, Box<dyn std::error::Error>> {
+        let mut svc = GrpcServer::new(self)
+            .max_decoding_message_size(config.max_decoding_message_size)
+            .max_encoding_message_size(config.max_encoding_message_size);
+
+        if let Some(encoding) = &config.accept_compressed {
+            if encoding.eq_ignore_ascii_case("Gzip") {
+                svc = svc.accept_compressed(CompressionEncoding::Gzip);
+            };
+        }
+
+        if let Some(encoding) = &config.accept_compressed {
+            if encoding.eq_ignore_ascii_case("Gzip") {
+                svc = svc.send_compressed(CompressionEncoding::Gzip);
```

Comment:
problem here with the if on top, you override svc with the same check

Reply:
We have to. 

In general, the configuration is quite hard be cause of the use of https://cliffle.com/blog/rust-typestate/ in the builders

Reply:
we better use else if or other to be sure to allocate svc only once.

there is no need to do this : `svc = svc.accept_compressed(CompressionEncoding::Gzip);` 
if you do this after :  `svc = svc.send_compressed(CompressionEncoding::Gzip);`

Reply:
I got It, we'll improve this